### PR TITLE
Add reinitialize_charm to harness

### PR DESCRIPTION
**Pull request by sed-i**
This PR adds a convenience method to harness for simulating a charm upgrade ([example](https://github.com/canonical/prometheus-k8s-operator/blob/7ae9e8fe139323463bc40e68d91d96cde29c0caf/tests/unit/test_remote_write.py#L271)).

Manually emitting `upgrade_charm` could probably be enough to cover most of the need, but having all the correct events emitted in the correct order makes it easier to choose unit tests over integration tests where possible.

### CODE REVIEW

File: ops/testing.py [hunk 1]
```diff
@@ -187,6 +187,40 @@ class TestCharm(self._charm_cls):
...
+        if self._charm is None:
+            raise RuntimeError('cannot call the upgrade method before begin')
+
+        # Reinitialize the charm-under-test
+        self._charm.on.stop.emit()
+        self._framework = framework.Framework(
+            self._storage, self._charm_dir, self._meta, self._model)
+        self._charm = None
+
+        # self.framework._objects.pop(self._charm.__class__.__name__)
```
Comment (PietroPasotti) [inline ops/testing.py:208]: remove commented code

File: ops/testing.py [hunk 2]
```diff
@@ -187,6 +187,40 @@ class TestCharm(self._charm_cls):
...
+        This triggers stop, upgrade-charm, config-changed, leader_settings_changed and start hooks.
+        This does NOT trigger a pebble-ready hook (use :meth:`.container_pebble_ready` for that).
+
+        Before calling :meth:`upgrade`, a Charm instance must exist, so you must call
+        :meth:`.begin` or :meth:`.begin_with_initial_hooks` first.
+        """
+        if self._charm is None:
+            raise RuntimeError('cannot call the upgrade method before begin')
+
+        # Reinitialize the charm-under-test
```
Comment (PietroPasotti) [inline ops/testing.py:202]: I think `reinitialize_charm` should be method-worthy on its own (i.e. without emitting `stop` first, and without emitting any events after.

Use cases: https://github.com/PietroPasotti/traefik-route-k8s-operator/blob/main/tests/unit/conftest.py (see last def)

File: test/test_testing.py [hunk 3]
```diff
@@ -2396,6 +2396,145 @@ def test_begin_with_initial_hooks_install_sets_status(self):
...
+        harness = Harness(RecordingCharm, meta='''name: test-app''')
+        self.addCleanup(harness.cleanup)
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
+        harness.charm.changes = []  # Clean up records to have a fresh count for `upgrade`
+        harness.upgrade()
+        self.assertEqual(
+            harness.charm.changes,
+            [
+                # {'name': 'stop'},  # the "stop" record is dropped due to charm reinit on upgrade
```
Comment (PietroPasotti) [inline test/test_testing.py:2409]: can't we hold on to the charm and test?

```python
dying_charm = harness.charm
harness.upgrade()
self.assertEqual(dying_charm.changes[-1], ...)
```

File: test/test_testing.py [hunk 4]
```diff
@@ -2396,6 +2396,145 @@ def test_begin_with_initial_hooks_install_sets_status(self):
...
+                {'name': 'upgrade-charm'},
+                {'name': 'config-changed', 'data': {}},
+                {'name': 'start'},
+            ]
+        )
+
+    def test_upgrade_status_consistency(self):
```
Comment (PietroPasotti) [inline test/test_testing.py:2522]: Not sure why we're testing this. In general it's not true, is it? An upgrade could in principle introduce non-backwards compatible changes (e.g. add a mandatory config option...)

**Review by PietroPasotti — APPROVED** [review]
I like this a lot! Only minor comments.

**Comment by rwcarlsen** [comment]
This definitely looks useful.  However, particularly after discussions at our recent roadmap meeting, I think this sort of thing probably needs to start as a library.  Maybe we need to come up with an ops framework "boost" library that contains high-power experimental libs (that selectively get merged into the framework).  The issue here is that there are all sorts of event sequences a charm should be checking.  It especially gets complex when you account for things like interleaving processes - e.g. upgrade during scaling op, etc.  Or even simply - losing workload connectivity at various points in a sequence.  If there are several things charms want to test, then which ones do we want to promote as "special" for testing with the harness?  In general I think just adding more special "Harness.simulate_sequence_[foo]" methods will not necessarily be the best/scalable approach.  Creating tools/helpers to assist and guide charm testing along these lines is something we will want to improve and work on from the framework side.

**Comment by PietroPasotti** [comment]
> 

I can agree with that as well, but I think that the 'reinitialize_charm' method would be a generic utility worthy of making it directly into the harness. I've already written three utility functions for that same functionality in as many projects.

**Comment by rwcarlsen** [comment]
When I've needed to "reset" things for testing, I just create a new Harness instance.  Are there use cases y'all have run into where that doesn't work well?

**Comment by rwcarlsen** [comment]
`begin_with_initial_hooks`' existence itself is considered controversial by some ;-)    I'm more open to a generic reinit/reset a charm functionality.  I know G and M have expressed a general concern for expanding the corpus of testing/event sequences in begin_with_initial_hooks fashion.

File: test/test_testing.py (hunk 3 above)
Comment (sed-i) [reply test/test_testing.py:2409]: Done in next commit.

**Comment by sed-i** [comment]
> `begin_with_initial_hooks`' existence itself is considered controversial by some

I get that, but when a test can be a unit test or an integration test, I'd rather it be a unit tests, and that's where "generic event sequencing" really shines.

**Comment by sed-i** [comment]
Side note: a [hypothesis extension](https://hypothesis.readthedocs.io/en/latest/strategies.html) for harness could be very interesting for test scalability and coverage. Silly example:

```python
from ops.testing.strategy import happy_path
from hypothesis import given

@given(sequence=happy_path(...))
def test_startup(self, sequence):
    harness = Harness(MyCharm, sequence=sequence)
```

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
After some internal discussion, the conclusion is:
- we like and agree with the 'reinitialize charm' utility being present in harness.
- we don't think that it's scalable and future-proof to add specific sequence-of-events methods to harness. We think that, indeed, there should be a separate pipeline for that. Something like a `harness.simulate_events(SequenceOfEvents)` that can be combined with a hypothesis extension and possibly several built-in common-case sequences, such as 'initial hooks', 'update', 'pod churn', 'machine reboot' etc... This needs not be addressed in this pr, in fact it should be a separate one.

**Comment by sed-i** [comment]
Reduced scoped to reinitialize only.
This now feels a bit confusing:
- What does reinitialize mean outside the context of an upgrade?
- Should statuses, relations, relation data be retained or dropped across a re-init?

File: test/test_testing.py [hunk 5]
```diff
@@ -2396,6 +2396,77 @@ def test_begin_with_initial_hooks_install_sets_status(self):
...
+        self.addCleanup(harness.cleanup)
+        harness.set_leader()
+        harness.begin_with_initial_hooks()
+
+        harness.charm.unit.status = ActiveStatus("unit before reinit")
+        harness.charm.app.status = ActiveStatus("app before reinit")
+        harness.reinitialize_charm()
+        self.assertEqual(harness.charm.unit.status, ActiveStatus("unit before reinit"))
+        self.assertEqual(harness.charm.app.status, ActiveStatus("app before reinit"))
+
```
Comment (PietroPasotti) [inline test/test_testing.py:2469]: Might be good to add a test with some __init__ logic affecting the charm state, like I pseudocoded in the first comment.

**Review by PietroPasotti — APPROVED** [review]
I think the docstring needs some sharpening, but the rest looks good now. Thanks!

File: ops/testing.py [hunk 6]
```diff
@@ -187,6 +187,16 @@ class TestCharm(self._charm_cls):
...
 
+    def reinitialize_charm(self) -> None:
+        """Re-initialize the charm under test."""
+        if self._charm is None:
+            raise RuntimeError('cannot call the reinitialize_charm method before begin')
+
+        self._framework = framework.Framework(
+            self._storage, self._charm_dir, self._meta, self._model)
+        self._charm = None
+        self.begin()
```
Comment (rwcarlsen) [inline ops/testing.py:198]: Are we sure we want a begin() call here?

**Comment by rwcarlsen** [comment]
I just realized that this PR seems to be somewhat related to this issue: https://github.com/canonical/operator/issues/736

File: ops/testing.py (hunk 6 above)
Comment (PietroPasotti) [reply ops/testing.py:198]: Umm. Better than begin_with_initial_hooks, I guess...
This is a `reinitialize` method -> we want the charm to be reinitialized. 
We do also want the possibility to intervene before the charm's `__init__` is run (to simulate a relation, some config change, etc...). But we can't simulate those things unless there is a framework, I suppose.
Do we need to decouple the framework setup and the charm's init?

Comment (PietroPasotti) [reply ops/testing.py:198]: Or maybe we need not to reinit the framework, unless the user asks us to?
My approach before this feature was (is) to `forget()` each object that had a handle, then reinitialize the charm (and the charm only). The framework remained.

**Comment by sed-i** [comment]
Abandoned. Closing.

**Pull request closed without merging**